bokeh/bokeh
# Fix rendering on JupyterHub 4.1: respect `absolute_url` for protocol determination, support `data-absolute-url`

- [x] issues: fixes #14002
   - which together with https://github.com/holoviz/pyviz_comms/pull/125
   - helps to fix https://github.com/holoviz/panel/issues/7039
- [x] tests passed (see below for manual testing instructions)
- [ ] release document entry (if new feature or API change)

This PR:
-  fixes `ws`/`wss` determination when `absolute_url` is passed
- supports reading the absolute URL from `data-absolute-url` of an iframe as `srcdoc` iframes cannot pass it otherwise (e.g. there is no motion of URL parameters for these)

| Before | After |
|--|--|
| ![Screenshot from 2024-07-29 17-16-06](https://github.com/user-attachments/assets/c898db35-aacb-4601-b8e5-ac6b55db9be5) | ![Screenshot from 2024-07-29 17-29-59](https://github.com/user-attachments/assets/3a5d2057-e128-4aa0-95db-d16cef2b8a6b) |


<details>

<summary>How to test without installing JupyterHub</summary>

1. Install pyviz_comms with https://github.com/holoviz/pyviz_comms/pull/125 applied
2. Run `jupyter --paths` and choose a writable config path from the `config:` section, e.g. `/home/username/my-env/etc/jupyter`
3. Open (create if needed) a `/home/username/my-env/etc/jupyter/jupyter_server_config.py`, paste the following:
   ```py
    c.LabApp.tornado_settings = {
        'hub_prefix': '/hub',
        'hub_host': 'test',
        'user': 'test',
        'headers': {'Content-Security-Policy': "frame-ancestors 'none'"}
    }
   ```
4. Start `jupyter lab`
5. Open dev tools Network tab, click on a request and expand Response Headers, verify that CSP for frame ancestors is set to none:
   ![image](https://github.com/user-attachments/assets/f446bacb-5be2-4bef-9508-a4b844adb8bf)
6. Ensure that page data includes `hubPrefix` by viewing the source of JupyterLab page
   ![Screenshot from 2024-07-30 10-06-57](https://github.com/user-attachments/assets/cf74d7a2-a2e2-40bd-b876-7b294725047e)
7. Follow reproduction steps in https://github.com/holoviz/panel/issues/7039


</details>
